### ISSUE TITLE
Implement plugin system for commands

### ISSUE THREAD

**Issue by shimpeko** [radeklat/delfino#30]
### What
Implement a plugin system for Delfino commands. Which will let us manage commands in a repository separated from the Delfino's repo. It also means that users will be able to manage their own commands in one place (repository) as a package and install the package in multiple repositories.

With a plugin system, the package dependency config should look like this.

```
[tool.poetry.dependencies]
delfino = "^0.17"
delfino_commands = "^1.0"
my_delfino_commands = "^1.0"
```

### Why
Now this repository is the only centralised place to manage commands; which is not suitable to manage commands for uncommon/personal/organization-specific tasks. Also, managing Delfino's default commands as plugins open up a wider opportunity for decentralized development.

**Comment by shimpeko** [comment]
Here are some topics I'd like to discuss;

### Naming: "command", "plugin"

Currently, a "plugin" means a "click.Command". But I'd like to use a "plugin" to point to a "distribution" which provides "click.Command"s.

### Granuraty of "plugin".

A plugin (=distribution package) can be a command OR set of commands.

I suggest defining a plugin as a set of commands, which let us install a set of commands in one distribution package which can be installed by pip install.

My use case for the plugin system is managing commands for our company within an internal repository. With such a use case, I don't see any requirements for granular management, a plugin per command, and "a plugin as a set of commands" requires less administrative effort.


### How to handle command name collision

Two or more plugins can provide commands with the same name. For example, both `plugin_a` and `plugin_b` provide `typecheck` commands.


I don't have recommendation yet but here is the idea.

#### Maintain package load order

Manage load order/priority per plugin in the pyproject.toml. Defaults to random? but users can provide an ordered list or priory through the config file.

#### Prefix command name with distribution name

When both `plugin_a` and `plugin_b` provide `typecheck` commands, Delfino prefix the command with distribution name (=plugin name) as `plugin_a_typecheck` and `plugin_b_typecheck` when loading commands. 

It may be useful if we allow users to define alias per command via pyproject.toml in this case. 

#### Other

- Do not register commands from plugin packages. Let users to define custom commands (in their repo local) and forward them to a package/plugin they want to use.
- Maintain command <-> package/plugin name mapping
- use 'extra' (`plugin_package[command]`). it doesn't make much sense for the plugin package because pip installs plugin_package` does install nothing.

**Comment by shimpeko** [comment]
Initial implementation can be found here: https://github.com/radeklat/delfino/compare/main...shimpeko:delfino:sk/build-plugin-entry-point


**About naming: "command", "plugin"**; I haven't resolved this. Some code use "plugin" to point to a distribution package whereas some use it for a click.command. I think it make sense to use the word "command" to point to commands like typecheck etc.. and using "plugin" for a distribution package.

**About granuraty of "plugin (distribution package)**; I've defined a plugin as a *set of commands*.

**About handling name collision**; I've implemented collision handling with a configuration in pyproject.toml as https://github.com/radeklat/delfino/compare/main...shimpeko:delfino:sk/build-plugin-entry-point#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R151-R153. I've considered using `extra` but didn't take it as it is hard to enforce plugin package distribution to use it. 

I've chosen to use the "package metadata" to discover plugins (ref. https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/). This allows us to name a distribution package without restriction (e.g. `global-python-util`) and to use arbitrary package structure. But I don't have a strong opinion here.
